### PR TITLE
[Common, ffmpeg]Exp: Add a ffmpeg downloader and player for any URL

### PR DESCRIPTION
**Pull request by cnbeining**
I am hoping we can use FFmpeg to deal with HLS(for I do not want to write our own dynamic M3U parser). And preferably provide some hints of working with HDS.

So:

I am opening a new function called ```download_url_ffmpeg``` that can be called in extractors that would work just like ```download_rtmpdump_stream```. 

However, since FFmpeg is sooooooooooooo powerful, DO NOT use it to download everything!

@soimort Merge with caution. Not fully tested. PR raised mostly with faith.

---
This change is [Reviewable](https://reviewable.io/reviews/soimort/you-get/1237)

### CODE REVIEW

**Comment by soimort** [comment]
Rebase this PR on @zhangn1985's #1236 and let it handle the M3U8 thing, so we can at least see how well it works.

File: src/you_get/processor/ffmpeg.py [hunk 1]
```diff
@@ -199,3 +199,64 @@ def ffmpeg_concat_mp4_to_mp4(files, output='output.mp4'):
...
+        output = output_dir + output
+        
+    ffmpeg_params = []
+    #should these exist...
+    if len(params) > 0:
+        for k, v in params:
+            ffmpeg_params.append(k)
+            ffmpeg_params.append(v)
+        
+    print('Downloading streaming content with FFmpeg, press Ctrl+C to stop recording...')
```
Comment (soimort) [inline src/you_get/processor/ffmpeg.py]: If you press <kbd>Ctrl-C</kbd> during the ffmpeg demux process (MPEG-4), most likely all you get is a corrupted file. (I guess)

Comment (cnbeining) [inline src/you_get/processor/ffmpeg.py]: As long as you do not press that multiple times, the file should be good.......I guess.

**Comment by cnbeining** [comment]
And for the rebase work... how can I merge the PR?

My local head is from cnbeining/you-get ...

**Comment by soimort** [comment]
> My local head is from cnbeining/you-get ...

Assume you're on `ffmpeg-live` branch, I think this should work:

```
$ git checkout -b zhangn1985-fix-iqiyi ffmpeg-live
$ git pull https://github.com/zhangn1985/you-get.git fix-iqiyi
```

File: src/you_get/extractors/iqiyi.py [hunk 2]
```diff
@@ -137,9 +137,9 @@ def prepare(self, **kwargs):
         assert info['code'] == 'A00000', 'can\'t play this video'
 
         for stream in info['data']['vidl']:
-           stream_id = self.vd_2_id[stream['vd']]
-           stream_profile = self.vd_2_profile[stream['vd']]
-           self.streams[stream_id] = {'video_profile': stream_profile, 'container': 'm3u8', 'src': [stream['m3u']], 'size' : 0}
+            stream_id = self.vd_2_id[stream['vd']]
+            stream_profile = self.vd_2_profile[stream['vd']]
+            self.streams[stream_id] = {'video_profile': stream_profile, 'container': 'm3u8', 'src': [stream['m3u']], 'size' : 0}
```
Comment (zhangn1985) [inline src/you_get/extractors/iqiyi.py]: 没注意，竟然只有3空格。

**Comment by cnbeining** [comment]
```
  File "/home/travis/build/soimort/you-get/src/you_get/extractors/iqiyi.py", line 102
    vd_2_profile = {21: u'超清', 2: u'高清', 4: u'超高清', 17: u'全高清', 96: u'流畅', 1: u'标清'}
                                ^
SyntaxError: invalid syntax
make: *** [test] Error 1
The command "make test" exited with 2.
Done. Your build exited with 1.
```

我就不知道为什么Python3.2就死活不喜欢这句话了。

**Comment by cnbeining** [comment]
OK.

As tested in
https://github.com/cnbeining/you-get/tree/showroom : 


It is not so likely possible that the HLS file is corrupted. Normal M3U files are good, as long as you don't touch your Ctrl+C too soon.

So:

This PR should be good to merge. More tests welcomed.

**Comment by soimort** [comment]
#1236 has been merged, so there're some conflicts need to be resolved. (on your `iqiyi.py`)

**Comment by cnbeining** [comment]
I will just do a update of my iQiyi.

**Comment by cnbeining** [comment]
Conflict solved. Are we good to go?

**Comment by soimort** [comment]
My system just fucked up today and I have not tested anything yet. Plus, iqiyi.com is still giving me timeout as often.
I assume you've had enough time to test this PR yourself, so we're good.

File: src/you_get/extractors/iqiyi.py [hunk 3]
```diff
@@ -207,4 +246,4 @@ def extract(self, **kwargs):
 site = Iqiyi()
 download = site.download_by_url
 iqiyi_download_by_vid = site.download_by_vid
-download_playlist = site.download_playlist_by_url
+download_playlist = site.download_playlist_by_url
```
Comment (soimort) [inline src/you_get/extractors/iqiyi.py]: Again, do not change the permission of source files if you don't have to.

**Comment by gdgsdg123** [comment]
> Quote "[/src/you_get/processor/ffmpeg.py](https://github.com/soimort/you-get/blob/bf507e5e45e61223e898ab0d740fded90c35d69d/src/you_get/processor/ffmpeg.py#L258-L261)":
>
> ```Python
>     if FFMPEG == 'avconv':  #who cares?
>         ffmpeg_params += ['-c', 'copy']
>     else:
>         ffmpeg_params += ['-c', 'copy', '-bsf:a', 'aac_adtstoasc']
> ```
> (line 258 - 261)

It matters, for the very reason:

https://github.com/soimort/you-get/commit/fd93b0380d0c04c4766e520e53a6256437e39067#diff-16d8fe83f175530f99b68a216b2dc733

> ```ShellSession
> > you-get ""
>
> [DEBUG] get_content: https://www.google.com/search?tbm=vid&q=
> you-get: version 0.4.1456, a tiny downloader that scrapes the web.
> you-get: Namespace(URL=[''], auto_rename=False, cookies=None, debug=True, extractor_proxy=None, force=False, format=None, help=False, http_proxy=None, info=False, input_file=None, insecure=True, itag=None, json=False, no_caption=True, no_merge=False, no_proxy=False, output_dir='.', output_filename=None, password=None, player=None, playlist=False, skip_existing_file_size_check=False, socks_proxy=None, stream=None, timeout=600, url=False, version=False)
>
> ...
> ```
>
> ```ShellSession
> > you-get -F TD "https://www.iqiyi.com/v_19rr80hvpc.html"
>
> [DEBUG] get_response: https://www.iqiyi.com/v_19rr80hvpc.html
> [DEBUG] get_content: http://pcw-api.iqiyi.com/video/video/playervideoinfo?tvid=714550700
> [DEBUG] get_content: http://cache.m.iqiyi.com/tmts/714550700/714550700/?t=1602541641827&sc=e3856cb34ce988edbc4092aa7dbeeaaa&src=76f90cbd92f94a2e925d83e8ccd22cb7
> site:                爱奇艺 (Iqiyi)
> title:               我的前半生第1集
> stream:
>     - format:        TD
>       container:     m3u8
>       video-profile: 720p
>       m3u8_url:      http://cache.m.iqiyi.com/mus/205416101/d856056acdd45c5b0cb33b2563d97109/afbe8fd3d73448c9/0/20170726/19/af/915d43104b7ffa8adfb012d6db8492f1.m3u8?qd_originate=tmts_py&tvid=714550700&bossStatus=0&qd_vip=0&px=&src=3_31_312&prv=&previewType=&previewTime=&from=&qd_time=1602541644252&qd_p=df686102&qd_asc=1b73a2969d0cdecf7a2deeb45ad1323c&qypid=714550700_04022000001000000000_14&qd_k=e3856cb34ce988edbc4092aa7dbeeaaa&isdol=1&code=2&ff=dbts&iswb=0&qd_s=otv&vf=cd8a6fe1ddbc0bb53996f38e846f5596&np_tag=nginx_part_tag
>     # download-with: you-get --format=TD [URL]
>
> ffmpeg version 4.3.1 Copyright (c) 2000-2020 the FFmpeg developers
>   built with gcc 8.4.0 (OpenWrt GCC 8.4.0 r1275-8563c180)
>   configuration: --enable-cross-compile --cross-prefix=arm-openwrt-linux-gnueabi- --arch=arm --cpu=cortex-a9 --target-os=linux --prefix=/opt --pkg-config=pkg-config --enable-shared --enable-static --enable-pthreads --enable-zlib --disable-doc --disable-debug --disable-lzma --disable-vaapi --disable-vdpau --disable-outdevs --disable-altivec --disable-vsx --disable-power8 --disable-armv5te --disable-armv6 --disable-armv6t2 --disable-fast-unaligned --disable-runtime-cpudetect --enable-lto --disable-vfp --disable-neon --disable-x86asm --enable-gnutls --enable-libopus --enable-small --enable-libshine --enable-gpl --enable-libx264
>   libavutil      56. 51.100 / 56. 51.100
>   libavcodec     58. 91.100 / 58. 91.100
>   libavformat    58. 45.100 / 58. 45.100
>   libavdevice    58. 10.100 / 58. 10.100
>   libavfilter     7. 85.100 /  7. 85.100
>   libswscale      5.  7.100 /  5.  7.100
>   libswresample   3.  7.100 /  3.  7.100
>   libpostproc    55.  7.100 / 55.  7.100
> [hls @ 0xb35400] Skip ('#EXT-X-DISCONTINUITY')
>     Last message repeated 5 times
> [hls @ 0xb35400] Opening 'http://dx.data.video.iqiyi.com/videos/v0/20170726/19/af/8dda3abd618cdfc5b25e92197c7b7c82.dbts?qypid=714550700_04022000001000000000_14&start=0&end=1893160&contentlength=1893160&sd=1400&durr=9000&qdv=1&qd_uid=0&qd_tvid=714550700&qd_vip=0&qd_src=3_31_312&qd_tm=1602541644446&qd_ip=df686102&qd_p=df686102&qd_k=e3856cb34ce988edbc4092aa7dbeeaaa&ve=&sgti=&dfp=&qd_sc=8929dd48913e73fb5a23678d994fa80b' for reading
> [hls @ 0xb35400] Opening 'http://dx.data.video.iqiyi.com/videos/v0/20170726/19/af/8dda3abd618cdfc5b25e92197c7b7c82.dbts?qypid=714550700_04022000001000000000_14&start=1893160&end=3368772&contentlength=1475612&sd=10400&durr=6680&qdv=1&qd_uid=0&qd_tvid=714550700&qd_vip=0&qd_src=3_31_312&qd_tm=1602541644446&qd_ip=df686102&qd_p=df686102&qd_k=e3856cb34ce988edbc4092aa7dbeeaaa&ve=&sgti=&dfp=&qd_sc=8929dd48913e73fb5a23678d994fa80b' for reading
> Input #0, hls, from 'http://cache.m.iqiyi.com/mus/205416101/d856056acdd45c5b0cb33b2563d97109/afbe8fd3d73448c9/0/20170726/19/af/915d43104b7ffa8adfb012d6db8492f1.m3u8?qd_originate=tmts_py&tvid=714550700&bossStatus=0&qd_vip=0&px=&src=3_31_312&prv=&previewType=&previewTime=&from=&qd_time=1602541644252&qd_p=df686102&qd_asc=1b73a2969d0cdecf7a2deeb45ad1323c&qypid=714550700_04022000001000000000_14&qd_k=e3856cb34ce988edbc4092aa7dbeeaaa&isdol=1&code=2&ff=dbts&iswb=0&qd_s=otv&vf=cd8a6fe1ddbc0bb53996f38e846f5596&np_tag=nginx_part_tag':
>   Duration: 00:44:40.00, start: 1.400000, bitrate: N/A
>   Program 0
>     Metadata:
>       variant_bitrate : 0
>     Stream #0:0: Video: h264 ([27][0][0][0] / 0x001B), yuv420p(tv, bt709), 1280x720, 25 fps, 25 tbr, 90k tbn, 50 tbc
>     Metadata:
>       variant_bitrate : 0
>     Stream #0:1: Audio: eac3 ([135][0][0][0] / 0x0087), 48000 Hz, stereo, fltp, 96 kb/s
>     Metadata:
>       variant_bitrate : 0
> [aac_adtstoasc @ 0xb70620] Codec 'eac3' (86056) is not supported by the bitstream filter 'aac_adtstoasc'. Supported codecs are: aac (86018)
> Error initializing bitstream filter: aac_adtstoasc
> Stream mapping:
>   Stream #0:0 -> #0:0 (copy)
>   Stream #0:1 -> #0:1 (copy)
>     Last message repeated 1 times
> Downloading streaming content with FFmpeg, press q to stop recording...
> ffmpeg -y -i http://cache.m.iqiyi.com/mus/205416101/d856056acdd45c5b0cb33b2563d97109/afbe8fd3d73448c9/0/20170726/19/af/915d43104b7ffa8adfb012d6db8492f1.m3u8?qd_originate=tmts_py&tvid=714550700&bossStatus=0&qd_vip=0&px=&src=3_31_312&prv=&previewType=&previewTime=&from=&qd_time=1602541644252&qd_p=df686102&qd_asc=1b73a2969d0cdecf7a2deeb45ad1323c&qypid=714550700_04022000001000000000_14&qd_k=e3856cb34ce988edbc4092aa7dbeeaaa&isdol=1&code=2&ff=dbts&iswb=0&qd_s=otv&vf=cd8a6fe1ddbc0bb53996f38e846f5596&np_tag=nginx_part_tag -c copy -bsf:a aac_adtstoasc -- 我的前半生第1集.mp4
> Skipping captions.
> ```

**Pull request merged**